[http] HTTP Binding initial contribution

kelnos opened the pull request:
This is inspired by the old OpenHAB1 HTTP binding, but allows the user to create Things that make HTTP requests to fetch and update state, and allows for a linked item (of any ItemType) to display/control that state.

I was bummed that using the old http1 binding didn't allow me to easily create things/items through PaperUI, and decided it could use an update.  Initially I was planning on basing this on the old binding, but I decided to just write it from scratch since 1 vs. 2 concepts seem a bit different, and I decided to take a slightly different approach in configuring Things created by the binding.

One thing I haven't tested is the transformation function support for HTTP request/response bodies, but I wanted to get the ball rolling with this PR, as it's my first contribution here and I want to make sure I'm doing the right things.

Prebuilt binding: https://files.spurint.org/org.openhab.binding.http-2.4.0-SNAPSHOT.jar

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/handler/HttpThingHandler.java:
```diff
@@ -0,0 +1,405 @@
+
+    @Override
+    public void handleConfigurationUpdate(final Map<String, Object> configurationParameters) {
+        super.handleConfigurationUpdate(configurationParameters);
```
I recommend not to handle this yourself. Have a look at the super implementation. The Handler will be disposed/initialized if the user changes a configuration. And that is perfect for your case, you just parse the config once, in initialize.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/handler/HttpThingHandler.java:
```diff
@@ -0,0 +1,405 @@
+
+    @Override
+    public void handleRemoval() {
+        cancelStateFetch();
```
You don't need to handle removal. Dispose will be called as well.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/handler/HttpThingHandler.java:
```diff
@@ -0,0 +1,405 @@
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        if (channelUID.equals(getThing().getChannel(CHANNEL_STATE).getUID())) {
+            this.linkedItems = this.itemChannelLinkRegistry.getLinkedItems(channelUID);
```
Nice idea, I see where you want to go with this. But please don't keep references in a handler to items. Use the method `isChannelLinked` or so (I can't recall exactly) in your `updateState` method.

davidgraeff commented:
Thank you very much for this binding. I almost had done this myself and I see a need for having this.
The code is in general of high quality, but you fell into some pitfalls, that every new binding developer falls into. 

Try to avoid (at every price) to keep references on items in Thing handlers. The abstraction chosen on purpose keeps those two worlds apart.

OSGI and the Eclipse smarthome framework gives you some guarantees. Your service will always be provided with required configuration parameters for example. No need to check anything.

kelnos commented:
Ok, I believe I've addressed the code comments.  Going to look into your suggestion around dynamic channels now and will take a look at the MQTT binding.  I like that design a lot better, and indeed when I was first building this it felt messy to have to create a different Thing Type for each Item Type.

kelnos commented:
All right, I made this use dynamic channels rather than a thing type per channel type.  I haven't completely tested it to my satisfaction, but an initial spot check looks good.  Definitely a usability improvement; thanks for the suggestion!

kelnos commented:
Hmm, one thing I noticed while testing is that the defaults specified in the configuration aren't pre-populated when adding a new channel (like they were when I had the configuration on the Thing itself).  Is this a known limitation/issue, or is there a different way of specifying defaults for channel configuration?

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java:
```diff
@@ -0,0 +1,275 @@
+@NonNullByDefault
+public class HttpChannelState implements AutoCloseable {
+    private static State stateFromString(final String stateStr) {
+        final String stateStrTrimmed = stateStr.trim();
```
I did the same in MQTT, before someone told me about the `TypeParser` class. It offers static methods to convert a string to a State or Command.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpHandlerFactory.java:
```diff
@@ -0,0 +1,113 @@
+        this.itemChannelLinkRegistry = null;
+    }
+
+    private <K, V> Map<K, V> dictionaryToMap(final Dictionary<K, V> dict) {
```
You can remove this method when you use the extended `activate` method :)

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpHandlerFactory.java:
```diff
@@ -0,0 +1,113 @@
+        super.activate(componentContext);
+        final Map<String, Object> properties = dictionaryToMap(componentContext.getProperties());
+        this.config = new Configuration(properties).as(HttpHandlerFactoryConfig.class);
+        this.handlers.forEach((uid, handler) -> handler.updateBindingConfig(this.config.getConnectTimeout(), this.config.getRequestTimeout()));
```
I see, you hand the config to your handlers. Why not make it a handler configuration in the first place and save yourself those round-trips. Because you have not defined a `modified` method
```
 @Modified
  public void modified(Map<String, Object> params)
```
OSGI will also terminate (`deactivate`) your factory (including all its handlers) on a configuration change and call `activate` again. You really might be better of having this Thing/Channel configurations.

kelnos replied:
Yeah that makes sense...  moved them to channel config, since people might need different timeouts depending on server/endpoint/etc.  Then I won't need this at all in the factory and can delete some code...

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpHandlerFactory.java:
```diff
@@ -0,0 +1,113 @@
+public class HttpHandlerFactory extends BaseThingHandlerFactory {
+    private final Map<ThingUID, HttpThingHandler> handlers = new HashMap<>();
+
+    private HttpClient httpClient;
```
The static code checker will not like this line. The reason is `@NonNullByDefault`. That annotation enforces that every field is either annotated with `@Nullable` (which in our case is only partly correct, because OSGI will in every case set this for us), or the field is set in the constructor, so that it can not be null.

To tell the code checker that you yourself will care about this not being null, annotate it with `@NonNullByDefault({})` like in:
```
private @NonNullByDefault({}) HttpClient httpClient;
```

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/model/ErrorListener.java:
```diff
@@ -0,0 +1,22 @@
+@FunctionalInterface
+public interface ErrorListener {
+    void accept(ChannelUID channelUID, ThingStatusDetail errorDetail, String errorDescription);
+}
```
new line missing

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/model/HttpChannelConfig.java:
```diff
@@ -0,0 +1,168 @@
+    /**
+     * A class describing configuration for the HTTP request to make when fetching {@link State}.
+     *
+     * @author Brian J. Tarricone
```
I know that eclipse is auto suggesting this, but inner classes do not have the author tag according to the openHAB coding guidelines.

davidgraeff commented:
> All right, I made this use dynamic channels rather than a thing type per channel type. I haven't completely tested it to my satisfaction, but an initial spot check looks good. Definitely a usability improvement; thanks for the suggestion!

A small test bundle will make sure that it works. I will make a PR on the weekend, if you don't find time to add it yourself.

> Hmm, one thing I noticed while testing is that the defaults specified in the configuration aren't pre-populated when adding a new channel 

This technology is rather new (got introduced this year, I believe) and only used by two bindings so far. This sounds like a core issue. In MQTT I had the xml defined defaults repeated in my java configuration classes, so I never noticed this.

davidgraeff commented on addons/binding/org.openhab.binding.http/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,85 @@
+            <default>POST</default>
+            <options>
+                <option value="POST">POST</option>
+                <option value="GET">GET</option>
```
Allow PUT as well?

davidgraeff commented on addons/binding/org.openhab.binding.http/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,85 @@
+        </parameter>
+    </config-description>
+
+    <config-description uri="thing-type:http:config-all">
```
http basic auth would be awesome, but can be added in a follow up PR as well.

kelnos commented:
Ok, addressed all comments, including adding support for basic auth.  Haven't written a test bundle yet, but will look into that if you don't get to it before I do ;)

kelnos commented:
> This technology is rather new (got introduced this year, I believe) and only used by two bindings so far. This sounds like a core issue. In MQTT I had the xml defined defaults repeated in my java configuration classes, so I never noticed this.

Any gut-feeling guess as to whether the issue is in PaperUI, openHAB, or the eclipse-smarthome stuff?

davidgraeff commented:
Eclipse Smarthome is the faulty one in that case.

davidgraeff commented:
@openhab/2-x-add-ons-maintainers Could someone please retrigger jenkins

davidgraeff commented on addons/binding/org.openhab.binding.http/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,22 @@
+			<description>HTTP request timeout, in milliseconds.</description>
+			<default>5000</default>
+		</parameter>
+	</config-description>
```
I think those can be removed now?

davidgraeff commented on addons/binding/org.openhab.binding.http/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Discovery
+
+Automatic discovery is not supported.
```
Please remove this section.

davidgraeff commented on addons/binding/org.openhab.binding.http/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Binding Configuration
+
+The binding itself has no configuration.
```
Please remove this section.

davidgraeff commented on addons/binding/org.openhab.binding.http/README.md:
```diff
@@ -0,0 +1,128 @@
+
+The Thing itself does not support any configuration.  However, to be useful,
+any number of channels can be added to a created Thing.  Each channel can be
+configured with different URLs to fetch and update state.
```
Please remove this section. You can join "Each channel can be configured with different URLs to fetch and update state" with the introduction sentence of the next section, if you want to.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java:
```diff
@@ -0,0 +1,264 @@
+        } else {
+            try {
+                return TypeParser.parseState(STATE_TYPES, doTransform(transform, response.asString()));
+            } catch (final IllegalArgumentException e) {
```
If you handle the exception of `doTransform` anyway, why not use the original TransformationException, instead of wrapping it in a IllegalArgumentException like you did. You can handle both exception types here with:
`IllegalArgumentException|TransformationException e`

kelnos replied:
(I think you put this comment on the wrong line, but I see what you're getting at.)

So the issue is that `TransformationException` is a checked exception, so if I didn't transform it, I'd be unable to call it within a lambda in `doTransform()` to allow the exception to bubble out.  I agree that it makes sense to keep it separate, though, so instead of mapping on the Optional, I'll just do  isPresent+get instead and remove the wrapping.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/handler/HttpThingHandler.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    protected void updateState(final ChannelUID channelUID, final State state) {
+        final boolean needsUpdate = this.itemChannelLinkRegistry.getLinkedItems(channelUID).stream()
+                .anyMatch(item -> !state.equals(item.getState()));
```
That is now more efficient that the solution before. But how about using
```BaseThingHandler.isLinked```? You could remove `itemChannelLinkRegistry` completely.

kelnos replied:
So what I'm trying to accomplish here is to avoid `ItemStateEvent`s when the item's state hasn't actually changed.  Since we have to poll the HTTP URL for changes, each poll might fetch the same data as last time.  `isLinked` will of course tell me if the channel is linked to some item, but won't tell me the current state of that item and if the new state we've just fetched is actually different.

Initially I thought that the state can't change in a way such that we wouldn't know about it -- could just cache the last `Command` sent to the Thing, and then when we fetch state, compare it to that last command.  But, for example, for `INCREASE` and `DECREASE` commands, we don't know what the new state is after the HTTP server has processed the command, so on the next state fetch we don't know if the old state is valid or not.  We also don't necessarily know what to do if the command URL returns a failure status code: revert to the old state?  Change to some other unknown state?

I was thinking that the "protocol" I've made up between the binding and the HTTP servers might require that the Command URL return the new state of the device on success, but I expect that might be too much of a burden on existing HTTP services that might just not do that, and might not be modifiable.

But I guess there's a way to get most of the way there:

1. Cache the returned state after each request to the State URL.  Don't update state if the new state is the same as the last returned state.
2. Cache the last command sent after, if possible, converting it to an expected state.  If the last command sent isn't convertible to state (e.g. `INCREASE`), just clear out the cached state since it's now unknown.
3. After sending a command, immediately also fetch state (ignoring the usual refresh interval) in order to ensure our view of the world is correct.  (Maybe only do this for when we can't convert a command into a state.)

I'll think about this a bit more... will address other comments and come back to this later.

davidgraeff replied:
Most bindings just update channels with the same value. And the framework actually keeps an eye on the value, because you either get item changed or item updated events on the event bus. I would just push the value to the channel and use isLinked to not do any http request if not necessary.

kelnos replied:
Ah, didn't realize there was a difference between changed and updated.  Removing this check will simplify things quite a bit!  I'll also remove the change in `updateStatus()` for the Thing status updates.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/HttpChannelConfig.java:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    @SuppressWarnings("unused")
+    private @Nullable String stateUrl;
```
If you ever need to suppress the unused warning, something is wrong in your code.
And the static analyser is right here, because you assign to the "url" field and you are using the "stateurl" and "commandurl" field. Please check all unused fields in this class.

davidgraeff replied:
Now that I re-read the class, I have realized the indentation level and that `url` and `stateurl`/`commandurl` are not in one class. I think it would help reviewing and understanding the code if you could move the two inner classes into their own files

kelnos replied:
I'll do the refactor (agree that the inner classes make it confusing), but the reason the warnings show up is because the static analyzer / compiler can't tell that `Configuration.as(Class<?> cls)` uses reflection to assign values to those fields.  In this case the warning is entirely spurious.  (It's actually not an "unused variable" warning; it's a "private field is never assigned" warning, which weirdly can be suppressed by suppressing the "unused" warning.)

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/Transform.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            return service.transform(this.pattern, value);
+        } catch (final TransformationException e) {
+            throw new IllegalArgumentException("Failed to transform value: " + e.getMessage(), e);
```
Don't wrap the original transformation exception if you don't need to. Just throw TransformationException.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/util/HttpUtil.java:
```diff
@@ -0,0 +1,247 @@
+ * @author Brian J. Tarricone - Initial contribution
+ */
+@NonNullByDefault
+public class HttpUtil {
```
I think this class is more complicated than it needs to be with the CompletionStateResponseListeners etc. Have a look at my class:
https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/netutils/AsyncHttpClient.java

Especially line 112. Just use `getContentAsString()` within a `BufferingResponseListener`.

kelnos replied:
Ah, wasn't aware of `BufferingResponseListener`!  Thanks, updating...

davidgraeff commented:
@openhab/2-x-add-ons-maintainers  Assuming that @kelnos or myself add a test bundle to this PR, can we have this for OH2.4?

martinvw commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/HttpChannelConfig.java:
```diff
@@ -0,0 +1,263 @@
+            try {
+                commandMethod = Method.valueOf(this.commandMethod);
+            } catch (final IllegalArgumentException e) {
+                throw new IllegalArgumentException("Invalid commandMethod", e);
```
Why catch and rethrow the same exception, can you create specific ones? Or not catch it?

kelnos replied:
In this case it's to add more information that can be reported to the user about what they need to fix to clear the error.

martinvw commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/HttpChannelConfig.java:
```diff
@@ -0,0 +1,263 @@
+            try {
+                commandResponseTransform = Optional.ofNullable(this.commandResponseTransform)
+                        .map(s -> Transform.parse(bundleContext, s));
+            } catch (final IllegalArgumentException e) {
```
This IMHO now how exception handling should be used in Java, every line of code in this method has its own exception handling. The idea was that you could group and that it does not mix with you logic flow.

martinvw replied:
Okay, let's take one step back. 

I think we can safely stop throwing `IllegalArgumentException` (3x in Transform && 6x in this class), and create specific ones there. 

And if we can catch one then throw back something else.

It could then look like:

```java
public Optional<CommandRequest> getCommandRequest(final BundleContext bundleContext) throws IllegalArgumentException {
    return Optional.ofNullable(this.commandUrl).map(commandUrl -> {
        Optional<Transform> commandRequestTransform = Optional.ofNullable(this.commandRequestTransform)
                .map(s -> Transform.parse(bundleContext, s));

        Optional<Transform> commandResponseTransform = Optional.ofNullable(this.commandResponseTransform)
                .map(s -> Transform.parse(bundleContext, s));

        try {
            return new CommandRequest(
                    getCommandMethod(),
                    new URL(commandUrl),
                    Optional.ofNullable(commandUsername),
                    Optional.ofNullable(commandPassword),
                    Duration.ofMillis(this.commandConnectTimeout),
                    Duration.ofMillis(this.commandRequestTimeout),
                    this.commandContentType,
                    commandRequestTransform,
                    commandResponseTransform
            );
        } catch (MalformedURLException e) {
            throw new InvalidConfigurationException(String.format("Invalid value '%s' for commandUrl  %s", commandUrl, e.getMessage()));
        }
    });
}

private Method getCommandMethod() {
    try {
        return Method.valueOf(this.commandMethod);
    } catch (IllegalArgumentException e) {
        throw new InvalidConfigurationException(String.format("Invalid value '%s' for commandMethod", this.commandMethod), e);
    }
}
```

I think the source of this problem is that we have a config which is not in the enum types and not yet validated and then immediately go from the 'broken' model to some request. 

Maybe it would even make more sense to separate validation of the model from actual construction of actions. That is also how other bindings tend to do this. 

Because we should also provide feedback about this incorrect config to the user via the Thing-status, not just through an exception in a log.

kelnos replied:
I like your new suggestion, but it still causes the loss of one bit of information: if one of the transformations specified is invalid, this doesn't let us tell the user which one.

I just pushed a change that more or less does what you're asking; let me know what you think.

> Maybe it would even make more sense to separate validation of the model from actual construction of actions. That is also how other bindings tend to do this.

Is there a binding that you think does this particularly well that I could use as an example?

> Because we should also provide feedback about this incorrect config to the user via the Thing-status, not just through an exception in a log.

Sure, that's exactly why I've done it this way: these methods get called in the thing handler's `initialize()`, which catches the exception and updates the Thing status appropriately with a detailed message describing which configuration entry was invalid and why.

davidgraeff replied:
The configuration class is capable of parsing textual values and assign the corresponding enum value to a field, and will throw on invalid user input.

kelnos replied:
Ah, great, can at least eliminate the need to parse the HTTP method manually.

martinvw commented:
> @openhab/2-x-add-ons-maintainers Assuming that @kelnos or myself add a test bundle to this PR, can we have this for OH2.4?

@davidgraeff if @kaikreuzer approves the modelling we could maybe go for that. But there is very little time to actually test before the release. Maybe waiting one week would be the wise choice.

martinvw commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/HttpChannelConfig.java:
```diff
@@ -0,0 +1,263 @@
+        });
+    }
+
+    public Optional<CommandRequest> getCommandRequest(final BundleContext bundleContext) throws IllegalArgumentException {
```
I prefer to leave out `throws` for `RuntimeException` because it can also throw all other runtime-exceptions anyway.

kelnos replied:
I mainly put that there as a form of documentation, but can remove it.

kelnos commented:
Looks like there's an issue with the test runner (it's failed twice like this):
```
04:25:01 [ERROR] Reason: java.lang.RuntimeException: java.io.IOException: Could not create temp dir /tmp/tycho7231850332904423543equinox
04:25:01 [ERROR] while locating org.eclipse.tycho.core.DependencyResolver annotated with @com.google.inject.name.Named(value=p2)
```

davidgraeff commented:
@kelnos Don't take the jenkins build server too serious at the moment. It fails to download resources apparently. Travis succeed and that matters more.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java:
```diff
@@ -0,0 +1,246 @@
+            DecimalType.class,
+            DateTimeType.class,
+            StringType.class
+    );
```
This is the last actual error that I spotted. You are using a fixed list of STATE_TYPES.
But depending on your channel type you need slightly different lists.

For example the PointType has a higher priority then the HSBType in your list. But the point type will match any received string of the format "x,y,z", so the HSBType will never match and your color channel will end up with a PointType.

I'm afraid you need to define multiple lists (best do that in the Constants class), and depending on the channel type (channelTypeUID) you use the fitting list

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java:
```diff
@@ -0,0 +1,246 @@
+                            final Optional<CommandRequest> commandRequest,
+                            final BiConsumer<ChannelUID, State> stateUpdatedListener,
+                            final ErrorListener errorListener)
+    {
```
This constructor argument list is huge. Let's try to reduce it:
* A `HttpChannelState` is always tightly coupled to a handler, isn't it? Remove `httpClient`, `scheduler`, `errorListener`, `maxHttpResponseBodyLen` and replace it with the handler and grab all those values on demand from the handler.
* Provide the `Channel` instead of `channelUID` and `channelTypeUID` because you can deduce those two from a channel.

kelnos replied:
Hmm, not sure I agree with passing in the handler.  I generally prefer to avoid circular dependencies between classes, as it makes things easier to reason about.

kelnos replied:
I changed it to provide the `Channel` instead, at least.

kelnos replied:
Also removed `maxHttpResponseBodyLen` and had `HttpChannelState` figure it out itself.

davidgraeff replied:
If you want it clean you could extract necessary methods into an interface. But sticking to the separation of concerns pattern sometimes tangles objects together. In my opinion circular references are justified in that case.

You can as well have those two classes be merged, but you have split them on purpose. They still depend a lot on each other though, so why not go with the reference. @martinvw maybe you have an opinion on this?

kelnos replied:
I guess the two classes could be merged, but IMO that would make things pretty messy.

Right now the `HttpChannelState` constructor takes 8 arguments.  I wouldn't consider that a lot (though agreed that it isn't a small number either).  Passing the handler in instead would reduce it to 4.  Alternatively, creating an interface for the 3 callback lambdas and passing an instance of that instead would give us 6.

I'm just not seeing how a small improvement on a (fairly arbitrary, IMO) metric like "number of arguments to constructor" justifies adding a circular class dependency or putting things behind an extra layer of abstraction.  To me, it just makes the code harder to read and understand.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java:
```diff
@@ -0,0 +1,246 @@
+                        this.httpClient,
+                        method.name(),
+                        transformedUrl,
+                        commandRequest.getUsername(),
```
Could you maybe move username, password, connect timeout, request timeout into a super class ("Request") that CommandRequest and StateRequest inherit from? That way you could hand the Request object to `HttpUtil.makeRequest` to reduce this monstrosity of method signature.

davidgraeff commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/HttpChannelConfig.java:
```diff
@@ -0,0 +1,139 @@
+        });
+    }
+
+    private Optional<Transform> parseTransform(final BundleContext bundleContext,
```
I haven't found any problems anymore, but here is one big performance tip: It absolutely makes sense to cache the `TransformationService`. At the moment you are resolving this service for every http call (that uses transforms).

Imagine this: A user polls a rest interface every second with 10 different endpoint calls (doesn't sound unusual). So you are resolving this service 10 times a second.

And resolving a service is slllloooowwww. You are calling out to the OSGI framework and OSGI does a lot of string comparisons to finally find the service and hand it over, all thread safe and expensive of course.

Ideally just resolve it once per handler.

kelnos commented on addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/Transform.java:
```diff
@@ -0,0 +1,70 @@
+     * @return a new {@link Transform}
+     * @throws IllegalArgumentException if the provided string is invalid or if the transformation service is unavailable
+     */
+    static Transform parse(final BundleContext bundleContext, final String s) {
```
Ah, nah, that just gets called in those two places you noted, which only happens on configuration change.  Is it worth actually caching the instances somewhere else, so it doesn't even need to get looked up that often?

davidgraeff replied:
Nah once per config update is alright. I just haven't seen it often that the bundlecontext is carried over multiple method calls and I initially thought you are calling this more often.

kelnos commented:
Just bumped the versions to 2.5.0.  Any idea how soon we can get this merged?  Any remaining issues?

davidgraeff commented:
@kelnos Could you please squash your commits? (on the command line: `git rebase -i HEAD~43` and follow the instructions, then force push `git push -f`)
Because there are so many commits, they can't be reviewed individually anymore anyway.

And then I perform a final review together with @martinvw or @wborn.

kelnos commented on addons/binding/org.openhab.binding.http/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,143 @@
+		</parameter>
+	</config-description>
+
+	<config-description uri="thing-type:http:config-all">
```
Do you think it's not to inflexible to have a single URL for each thing?  One thing I was thinking is that a device (or whatever) might have different URLs for fetching state vs. sending commands.  Not sure how likely that is, but my feeling here is that this binding's main purpose is to deal with wacky things that don't fit into the standard openHAB way of doing things, so greater flexibility might be warranted.

davidgraeff replied:
The suggestion is because of complex data types like json/xml, I guess? Why didn't we go for this Bridge/Thing/Thing architecture in MQTT? Because MQTT topic payloads can be complex data types as well, exactly like http. We still went the route of Bridge/Thing/Dynamic channels with the cost of having more configuration per channel.
If we change the architecture here, the mqtt binding would need to be adopted as well, and my coap binding currently in development would use that architecture, too. But please, we should stay consistent for all those generic protocols.

kelnos replied:
To slightly backtrack on my last comment, I actually just realized that the example I have in the README actually becomes impossible with the proposal to move the URL config into the Thing.  (I guess you can technically create the same sitemap, but you can't have the "door switch" and "door camera" channels on the same Thing anymore.)

Would it make sense to have a sort of "master URL" on the Thing, and then allowing the channel config to override it?  That way, the common case of only needing one URL would be easy to configure.  Or is that too confusing from a UX perspective?

davidgraeff replied:
The Thing could be for the base-url part ("https://myserver.com/api/") with authentication+cache config, and a channel has the relative url part only ("sensor/door") and optional transformations.

davidgraeff replied:
> We'd probably require multiple relative paths (separate ones for sending & receiving) then, right?

I would make the sending URL optional and advanced, I think. Most http services are crud rest endpoints nowadays. Wdyt?

kelnos commented on addons/binding/org.openhab.binding.http/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,78 @@
+		<item-type>Image</item-type>
+		<label>Image</label>
+		<description>Current image to display</description>
+		<state readOnly="true"/>
```
How would this work framework-wise?  `RawType` is only a `State`, not a `Command`.

kaikreuzer replied:
Good point. Then better forget my comment 😇.

davidgraeff replied:
FYI: Images are read only channels in mqtt. I also do not see the use case and how a user would generate an image in the first place. 

A raw channel for all kinds of binary data could be introduced (also for mqtt, if I think about it), but not necessarily in this pr.

kelnos commented on addons/binding/org.openhab.binding.http/README.md:
```diff
@@ -0,0 +1,115 @@
+
+## Supported Things
+
+This binding supports a single thing type: `http`.
```
Yeah, I kinda struggled with this and then punted.  I like "endpoint", though.

kelnos commented:
@davidgraeff Yeah, was intending to squash before merging.  Decided to leave each "comment fixup" in a separate commit so it'd be easier to review just the updates if anyone cared to as we were going along.

@kaikreuzer Thanks for the feedback!  I'm a little concerned at this point, though: will the modeling changes you've suggested be good enough to merge?  I'm asking because this is a little frustrating: I've already done some large refactoring and model changes as a result of feedback, and your suggestions would involve moving some code _back_ where it originally was.  I've put quite a bit of time into this already, and I'd prefer not to be playing tug-of-war with reviewer comments.

kaikreuzer commented:
> I'm asking because this is a little frustrating: I've already done some large refactoring

Fully understood and sorry that you came into this situation.
The timing was a bit unfortunate: You created the PR shortly before the 2.4 release, when there wasn't time to reflect it in depth (thus I didn't comment at all then) - and the HTTP binding replacement is a HUGE one which we should not take lightly, but rather discuss all the different details before releasing it to users. @davidgraeff can tell how long it took from the first idea for MQTT to have it merged into 2.4 (and seeing all the user feedback and problems I wonder if even that might have been a bit premature)...

We (the "official" reviewers/maintainers) just recently discussed that we believe we need a two step review process - first check & discuss the skeleton/architecture/modelling with the contributor and then do in-depth code review as a second step. We will need to formalise and document that a bit. The idea is to ask contributors not to create a PR with all details only once they think they are finished, but rather create a WIP PR asap and discuss the general modelling/direction - this would have helped here to save efforts, I guess. So sorry again and thanks for all your work; I am very confident, it will result in something very useful :-) .

davidgraeff commented:
My 2ct regarding to user perception: Http2 will annoy one or the user, pretty sure about it. What I have learned is that a lot of users are still in the OH1 world and they have not yet confronted themselves with OH2 concepts.

But http is also the last huge oh1 binding (next to "expire") and users will adapt. A lot have successfully performed the migration for mqtt. Next time the "breaking changes" section should contain a full migration step to step tutorial though.

And yes it took a longer time to get the architecture for mqtt right and some framework additions and changes were required even.

kaikreuzer commented:
> next to "expire"

Off-topic: Before anyone starts with an expire2 binding: This should never ever have been a binding in the first place. This should be done as a framework feature instead.

kelnos commented:
First off, thanks @kaikreuzer for your explanation of your thoughts about the overall design/submission process -- really helps clarify things a bit.  I didn't realize the scope of what I was trying to do here!  I've spent the last ~7 years of my professional life frequently dealing with customer-facing API design, so I definitely get how important it is to get this as right as possible before putting something out in front of users and committing to supporting it.

Anyhow, let's see if I can gather the thoughts on this into something concrete.

Thing configuration:

* `baseUrl`, "Base URL" (required): URL that will be contacted to provide state or issue commands.
* `connectTimeout`, "Connect Timeout" (required, defaults to 3000ms): Time after which an incomplete TCP connect will cause the request to fail.
* `requestTimeout`, "Request Timeout" (required, defaults to 5000ms): Time after which the overall request fails if data has not been received from the server.
* `username`, "Username" (optional): Optional username to be provided via HTTP Basic Authentication.
* `password`, "Password" (optional): Optional password to be provided via HTTP Basic Authentication.
* `stateRefreshInterval`, "State Refresh Interval" (required, defaults to 60000ms): Time between state refreshes.
* `commandMethod`, "Command Method" (required, defaults to `POST`): HTTP Request method used when sending a command.
* `commandContentType` ,"Command Content Type" (required, defaults to `text/plain; charset=utf-8`): Content Type used when sending command requests.

Things will continue to support dynamic channels as in the current PR.

Channel configuration:

* `urlSuffix`, "URL Suffix" (optional): Optional suffix to append to the Thing's `baseUrl` when making HTTP requests for this channel.
* `stateResponseTransform`, "State Response Transform" (optional): Optional transformation function applied to the response received after fetching state.
* `commandRequestTransform`, "Command Request Transform (optional): Optional transformation function applied to the request when sending a command.
* `commandResponseTransform`, "Command Response Transform" (optional`): Optional transformation function applied to the response received after sending a command.

In addition, any configuration setting from the Thing can be overridden by providing it on a channel using advanced config.  On top of that, we can also allow on the channels, as advanced config, different values for state and command requests by prepending `state` or `command` to the config option (e.g. `stateUsername` allows overriding the username just for the state fetch, and `commandUrlSuffix` allows setting a different URL suffix for sending the command).

Implementation-wise, some notes:

* Within a Thing, we'll de-dupe all final state URLs to be fetched so that we don't make extra unneeded requests, and will pass (possibly transformed) values to each channel's state handler from the single response received.
* If some "duped" URLs have somewhat conflicting config (e.g., one has a request timeout of 5 seconds and another has that same timeout set to 10 seconds), we'll use the "most generous" config for the request we make (e.g. 10 seconds).
* Delaying URL validation until after the final URLs are constructed even allows utterly crazy things like setting `baseUrl` to `http` and then `stateUrlSuffix` to `://foo.com/whatever` and `commandUrlSuffix` to `s://bar.com/moo`, so you can send the state request using http and the command request using https, both to different servers, even.  I might give someone some serious side-eye for doing this, but it doesn't hurt usability to allow it.

Thoughts?

kaikreuzer commented:
Many thanks for your intense thoughts on this @kelnos! Sound a bit like a Swiss army knife to me 😎 . My only concern would be that there's the risk that it could become too complex from UX pov or also for implementing it. But that's up to you - if you feel ready for it, have a go, but I wouldn't mind if you decided to strip down some features for the start (like overwriting thing configurations on channels).

A few comments:
- Parameters that have a default (such as `connectTimeout` and others) should be marked optional, not mandatory. While they are mandatory for the implementation, they are optional for the user to provide (the implementation will automatically receive the default value, if the user didn't provide anything else).
- You say "we can't really allow %s in the URL when fetching state", but the http1 binding [says it uses java.util.Formatter`](https://www.openhab.org/addons/bindings/http1/#dynamic-urls) for dynamic urls, so that a simple "%%s" should result in a "%s" for the final url. Additionally, the http1 binding has a `format` parameter, which can be used to disable the processing of the url completely.
- `commandMethod` might better be called `httpVerb` or `httpMethod`(as in http1), wdyt?
- `commandContentType`: Is this really required? http1 says it always uses `text/plain` and that seemed to be sufficient. If the remote server expects more complex data, I'd rather think it is time for a dedicated binding for that service.
- I cannot see any possibility to post content (so why specify a content-type at all?) in a HTTP POST request (as it is possible in http1).
- The "out" feature of http1 seems to be lacking, i.e. specifying a command like `ON` (or special ones like `*` and `CHANGED`) upon which the outgoing HTTP request is initiated.
- Likewise, the possibility to [specify optional HTTP headers](https://www.openhab.org/addons/bindings/http1/#http-headers) to send with outgoing requests seems to be missing.

davidgraeff commented:
My 2cts:
* `commandContentType`: More and more webservers behave very strict on this header. And I can imagine that some users want to "post" json via a rule, hardcorded or via a new rule engine action
* The "out" feature: Can that not be replicated with rules, profiles and whatnotever?

kaikreuzer commented:
@davidgraeff:
1. Yes, but only valid if the command is actually sent as the content.
2. You can surely use those items then only in rules, but that's not really the point of items. In any case, it would be a huge step back in UX from http1 and there won't be any proper migration path. Safest way to keep people using http1 😉.

kelnos commented:
Hey, sorry for the disappearing act.  Real Life[tm] has been busy for me lately.

> `commandContentType`: Is this really required? http1 says it always uses text/plain and that seemed to be sufficient. If the remote server expects more complex data, I'd rather think it is time for a dedicated binding for that service.

I still think content-type is required to expose for sending commands.  With transformations, the plaintext command can be converted to a part of a JSON, XML, etc. document, and many servers (especially those that are aiming to be RESTful) do actually care that the content-type specified matches what it expects.  I don't think we can or should expect people to build dedicated bindings for services that work via HTTP, just because those services expect content other than plain text.  If that's the case, why offer the ability to transform the content at all in the first place?

> I cannot see any possibility to post content (so why specify a content-type at all?) in a HTTP POST request (as it is possible in http1).

The post content is the string of the command (optionally transformed).  I do note that the optional `postcontent` bit is missing in this reimagining of the binding; I personally didn't need it and so Ieft it out.  It could be added without much difficulty, though.  Curious if it seems important enough to be a first-class citizen in the UI, or if it should be advanced.

> Likewise, the possibility to specify optional HTTP headers to send with outgoing requests seems to be missing.

Hadn't thought about custom headers, but that shouldn't be too hard to add.  Agree that it's useful and would be required for back-compat with http1 anyway.

> The "out" feature of http1 seems to be lacking, i.e. specifying a command like ON (or special ones like `*` and `CHANGED`) upon which the outgoing HTTP request is initiated.

I guess you could say my binding only supports `*`.  I left out flexibility there in part because I didn't need it, but mainly because I couldn't think of a good way to make this work in the configuration UI.  Unless I'm missing something, there's no concept of "dynamic configuration", right?

Essentially I could see this being done in a similar way as dynamic channels: you have a single channel, but then in the configuration, you have a plus button that lets you add an arbitrary number of Command Configuration sections for each different command you need to support in a non-generic way.  But I don't think we can do that?  So I'm not sure how to model this from a configuration standpoint that's supportable in the UI.

kelnos commented:
Rebased off master and did a quick initial config reorg I discuss in https://github.com/openhab/openhab2-addons/pull/4342#issuecomment-450792804 (not tested yet aside from compilation).  This does not yet include the advanced config options, or the extra/missing features discussed later.

kelnos commented:
(Er, not sure why the "requested a revew from maintainers" thing popped up.  I did no such thing...)

J-N-K commented:
@kelnos, where are we here?

kelnos commented:
@J-N-K I've been a bit busy lately, but I think I'm also still looking for some more feedback as to what the right direction is. not too keen on going down another path without confirmation that what I'm doing will end up getting merged

boc-tothefuture commented:
Who can answer @kelnos question:
> I guess you could say my binding only supports *. I left out flexibility there in part because I didn't need it, but mainly because I couldn't think of a good way to make this work in the configuration UI. Unless I'm missing something, there's no concept of "dynamic configuration", right?

> Essentially I could see this being done in a similar way as dynamic channels: you have a single channel, but then in the configuration, you have a plus button that lets you add an arbitrary number of Command Configuration sections for each different command you need to support in a non-generic way. But I don't think we can do that? So I'm not sure how to model this from a configuration standpoint that's supportable in the UI.

Is it just @kaikreuzer? - Based on my read of the PR that info is needed for @kelnos to complete the binding.

martinvw commented:
@kaikreuzer some architectural direction of you is appreciated on this :-)

cpmeister commented:
What is the status of this?

kelnos commented:
@cpmeister please scroll up ever so slightly ;) https://github.com/openhab/openhab-addons/pull/4342#issuecomment-490294913

cpmeister commented:
@kelnos I'm just making sure that this Pr isn't dead 😄

kaikreuzer commented:
Oh hell, so you are all waiting since a year for some comment from me...? Sorry for that, you should have nudged me more often!

After all that time, I am not really in the details of the discussion anymore, so please forgive me, if I say any nonesense:

> Essentially I could see this being done in a similar way as dynamic channels
> you have a single channel, but then in the configuration, you have a plus button that lets you add an arbitrary number of Command Configuration sections

Hm, why trying to do that on a channel configuration level and not just allow multiple channels?
If you'd introduce a new config parameter like "commandFilter" (which defaults to "*" for the current behavior), the user would be able to add any number of channels and have a dedicated one for any specific command (like ON or whatever) he would like to see processed in a different way. Would that be an option?

udo1toni commented:
I may have missed something, but as far as I understood, there is nothing like the http1 cache in this successor. 
This feature helps keeping requests as low as possible, so I wonder if there is a chance to add a caching mechanism?

kaikreuzer commented:
@udo1toni I had mentioned the caching of the http1 binding in https://github.com/openhab/openhab-addons/pull/4342#discussion_r244545580 as well (in the context of binding multiple items to the same http endpoint), so I agree that the new binding needs to have caching in place as well. Haven't checked the code though and just hoping that @kelnos comes back for further discussion and finalizing this PR!

udo1toni commented:
Ah. missed that part. Thanks for clarification.

kaikreuzer commented:
Work continued on https://github.com/openhab/openhab-addons/pull/7851 - @kelnos, if you are still around, please feel free to join @J-N-K in that work, it isn't meant as a hostile take-over!

kelnos commented:
It's a shame it looks like they decided to start from scratch rather than pick up where I left off.  Considering that's the case, I don't think I'm interested in continuing.  Overall this has been a pretty demotivating experience.

kaikreuzer commented:
@kelnos I think @J-N-K only started #7851 because he believed that this PR here is stale. After all, there was no sign of live from you since my [comment on March 12](https://github.com/openhab/openhab-addons/pull/4342#issuecomment-598423950).
If you are still working on this binding and we can finish it in the coming weeks, I am sure that @J-N-K won't go through the effort to do a new implementation.

J-N-K commented:
Well. Then finish this one.

kaikreuzer commented:
Ok, since there's no further comment, let's push the work forward on #7851 instead.
